# Project 1 - CS C121

**Goal: predict the secondary structure of proteins from their amino
acid sequences.**

Dataset format: tsv file with the following columns:
- ID: identifier for the protein, the three-letter abbreviation for the amino acid, and the position of the amino acid within the protein.
- structure: the secondary structure label 
    - H = α-helix
    - B = residue in isolated β-bridge
    - E = extended strand, participates in β ladder
    - G = 310-helix
    - I = π-helix
    - P = κ-helix (poly-proline II helix)
    - T = hydrogen-bonded turn
    - S = bend
    - . = None

Files:
- train.tsv: contains id and the true secondary_structure label.
- test.tsv: contains only id — you'll predict secondary structures for this.
- sequences.fasta: full protein sequences for all proteins (IDs match those in train.tsv and test.tsv).

Implementation Details:
1. Preprocess Data: parse `sequences.fasta` to map protein IDs to sequences. Then align residue info in `train.tsv` to full sequences. (useful for ESM model)
2. Choose Feature Represntation: use ESM embeddings
3. Build the model: neural net trained on ESM embeddings 
4. Create a validation set from `train.tsv` and use it to evaluate accuracy before applying it to `test.tsv`
5. Make predictions on `test.tsv` with an added prediction column
6. Zip and submit to codabench
